File: frontend/src/lib/budget.ts
import { api } from './api';

export interface Budget {
  agent_id: string;
  base_budget: number;
  trust_modifier: number;
  risk_modifier: number;
  criticality: number;
  effective_budget: number;
  spent: number;
  remaining: number;
  utilization_pct: number;
  window: string;
  is_frozen: boolean;
  violation_count: number;
  department: string;
  currency: string;
}

export interface BudgetListResponse {
  budgets: Budget[];
  total: number;
  page: number;
  page_size: number;
}

export interface BudgetOverview {
  total_budget: number;
  total_spent: number;
  total_remaining: number;
  total_frozen: number;
  total_violations: number;
  department_breakdown: Array<{
    department: string;
    allocated: number;
    spent: number;
    utilization: number;
  }>;
  top_spenders: Array<{
    agent_id: string;
    department: string;
    spent: number;
    effective_budget: number;
    utilization_pct: number;
  }>;
}

export interface BudgetForecast {
  agent_id: string;
  current_spent: number;
  current_remaining: number;
  projected_spend: number;
  projected_remaining: number;
  days_remaining_in_window: number;
  daily_average_spend: number;
  will_exceed: boolean;
  confidence: number;
}

export const budgetApi = {
  listBudgets: async (params?: {
    page?: number;
    page_size?: number;
    department?: string;
    is_frozen?: boolean;
  }) => {
    const response = await api.get<BudgetListResponse>('/budgets', { params });
    return response.data;
  },

  getOverview: async () => {
    const response = await api.get<BudgetOverview>('/budgets/overview');
    return response.data;
  },

  getAgentBudget: async (agentId: string) => {
    const response = await api.get<Budget>(`/budgets/agents/${agentId}`);
    return response.data;
  },

  getAgentForecast: async (agentId: string) => {
    const response = await api.get<BudgetForecast>(`/budgets/agents/${agentId}/forecast`);
    return response.data;
  },

  getFormula: async () => {
    const response = await api.get<any>('/budgets/formula');
    return response.data;
  },
};

File: backend/app/schemas/budget.py
"""CAZZ SHIELD — Budget Schemas"""
from pydantic import BaseModel
from typing import Optional
from datetime import datetime


class BudgetResponse(BaseModel):
    agent_id: str
    base_budget: float
    trust_modifier: float
    risk_modifier: float
    criticality: float
    effective_budget: float
    spent: float
    remaining: float
    utilization_pct: float
    window: str
    is_frozen: bool
    violation_count: int
    department: Optional[str] = None
    currency: str

    class Config:
        from_attributes = True


class BudgetListResponse(BaseModel):
    budgets: list[BudgetResponse]
    total: int
    total_allocated: float
    total_spent: float
    total_remaining: float


class BudgetTransactionResponse(BaseModel):
    id: str
    agent_id: str
    transaction_type: str
    amount: float
    balance_before: float
    balance_after: float
    description: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class BudgetForecastResponse(BaseModel):
    agent_id: str
    current_spent: float
    current_remaining: float
    projected_spend: float
    projected_remaining: float
    days_remaining_in_window: int
    daily_average_spend: float
    will_exceed: bool
    confidence: float


class BudgetOverviewResponse(BaseModel):
    total_budget: float
    total_spent: float
    total_remaining: float
    total_frozen: int
    total_violations: int
    department_breakdown: list[dict]
    top_spenders: list[dict]
